lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,259 @@
+ */
+'use strict';
+
+/* globals self */
+
+/** @typedef {import('./dom.js')} DOM */
+/** @typedef {LH.Artifacts.Rect} Rect */
+/** @typedef {{width: number, height: number}} Size */
+
+/**
+ * @param {number} value
+ * @param {number} min
+ * @param {number} max
+ */
+function clamp(value, min, max) {
+  if (value < min) return min;
+  if (value > max) return max;
+  return value;
+}
+
+/**
+ * @param {Rect} rect
+ */
+function getRectCenterPoint(rect) {
+  return {
+    x: rect.left + rect.width / 2,
+    y: rect.top + rect.height / 2,
+  };
+}
+
+class ElementScreenshotRenderer {
+  /**
+   * Given the location of an element and the sizes of the preview and screenshot,
+   * compute the absolute positions (in screenshot coordinate scale) of the screenshot content
+   * and the highlighted rect around the element.
+   * @param {Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   * @param {Size} screenshotSize
+   */
+  static getScreenshotPositions(
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords, screenshotSize) {
+    //
+    const elementRectCenter = getRectCenterPoint(elementRectInScreenshotCoords);
+
+    // Try to center clipped region.
+    const screenshotLeftVisibleEdge = clamp(
+      elementRectCenter.x - elementPreviewSizeInScreenshotCoords.width / 2,
+      0, screenshotSize.width - elementPreviewSizeInScreenshotCoords.width
+    );
+    const screenshotTopVisisbleEdge = clamp(
+      elementRectCenter.y - elementPreviewSizeInScreenshotCoords.height / 2,
+      0, screenshotSize.height - elementPreviewSizeInScreenshotCoords.height
+    );
+
+    return {
+      screenshot: {
+        left: screenshotLeftVisibleEdge,
+        top: screenshotTopVisisbleEdge,
+      },
+      clip: {
+        left: elementRectInScreenshotCoords.left - screenshotLeftVisibleEdge,
+        top: elementRectInScreenshotCoords.top - screenshotTopVisisbleEdge,
+      },
+    };
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {HTMLElement} mask
+   * @param {{left: number, top: number}} positionClip
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   */
+  static renderClipPath(dom, mask, positionClip,
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords) {
+    // Normalize values between 0-1.
+    const top = positionClip.top / elementPreviewSizeInScreenshotCoords.height;
+    const bottom =
+      top + elementRectInScreenshotCoords.height / elementPreviewSizeInScreenshotCoords.height;
+    const left = positionClip.left / elementPreviewSizeInScreenshotCoords.width;
+    const right =
+      left + elementRectInScreenshotCoords.width / elementPreviewSizeInScreenshotCoords.width;
+
+    const clipId = `clip-${top}-${bottom}-${left}-${right}`;
+    const clipPathSvg = dom.createElement('div');
+    clipPathSvg.innerHTML = `<svg height="0" width="0">
```

Comment:
I know it's a pain, but we should never use `innerHTML`. Things start off safe but then abstractions develop that hide what's going on, then other contributors change how the abstractions are used, etc.

One approach would be the SVG could also live in the template and this method could fill in the `points` attributes (and `id`).

Reply:
I think we should defer the extra complexity until such abstractions develop and warrant it. As it is, it is 100% safe, yes? No way to inject arbitrary html based on the result of division operations.

Reply:
> I think we should defer the extra complexity until such abstractions develop and warrant it.

The issue is that often no one notices the abstractions or changes that warrant it (if they always did, then there would never be an issue :).

We should just never use `innerHTML` anywhere. We should probably have a linter rule for that (and it's likely we'd quickly get a rosie CL to fix it in the PSI frontend after it landed). I don't think the complexity is significantly different, matching what we do everywhere else in the report, and it's at least arguably better to have the markup with the other markup and not embedded in a JS string.

Reply:
ok, started on this.

except I hit a weird bug ... The svg clipPath element doesn't work unless I do this ugly hack ...

```
clipPathEl.innerHTML = clipPathEl.innerHTML;
```

Reply:
seems I doubled the trouble here :)

Reply:
weird, we have svg using url() references in other places (like the PWA category renderer) that clone the svg [from a template](https://github.com/GoogleChrome/lighthouse/blob/de072589d07d867403adb31728210829a4c943b5/lighthouse-core/report/html/templates.html#L647) without problem. Maybe it needs to be wrapped in an html element and not just `<template><svg>...</svg></template>` or something like that?

Reply:
oh, sorry, actually looked at the code now :) Maybe worth trying the template approach (and `setAttribute()`ing the points) to see if it has the same problem?

Reply:
absolutely don't get why using the `*NS` methods don't work ..

Reply:
got it. I had to not use `setAttributeNS` for the elements I created using `createElementNS`, for some reason.

so ... do we need to add `ns` as a property to dom.createElement??

Reply:
You can see what did/didn't work here: https://github.com/GoogleChrome/lighthouse/blob/d7cd541a6fd2e223f8599851964493df1be72ff9/lighthouse-core/report/html/renderer/element-screenshot-renderer.js#L81